### PR TITLE
complete todoList

### PR DESCRIPTION
**Pull request by Starkiller1988**
Free to review my code

Added my To-Do Component with functions

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import React, { Component } from "react";
+import React, { UseState } from "react";
```
Comment (mpagels) [inline src/App.jsx:1]: `UseState` starts with a small u -> `useState`
and: do you need that import here?

File: src/App.css [hunk 2]
```diff
@@ -157,50 +155,94 @@ border-radius: 10px;
...
+  font-family: Poetsen One;
+  font-size: 22px;
+}
+
+.back{
+  background-color: #F3E779;
+  height: 100%;
+}
+
+/* Calculator Component */
```
Comment (mpagels) [inline src/App.css:245]: Comment. Would recommend deleting it. You don't want commented code in production.

File: header&footer/footer/footer.jsx [hunk 3]
```diff
@@ -1,18 +1,17 @@
...
-            <li></li>
-            <li></li>
-
-        </ul>
+      <ul>
+        <li></li>
+        <li></li>
+        <li></li>
+        <li></li>
+      </ul>
```
Comment (mpagels) [inline header&footer/footer/footer.jsx:12]: Why is the list empty?

File: README.md [hunk 4]
```diff
@@ -1 +1,11 @@
 # hawify
```
Comment (macfrei) [inline README.md:1]: Nice job! Maybe add information about how to start your project if someone clones it?

File: components/calendar_components/Task.jsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      <label className="labeltasks">
+        <div>
+          <input
+            type="checkbox"
+            checked={task.complete}
+            onChange={handleTaskClick}
+            className="check"
+          />
+        </div>
+        <div className="task">{task.name}</div>
```
Comment (macfrei) [inline components/calendar_components/Task.jsx:19]: Do you need the `div`s in here? 🤔

File: components/calendar_components/TaskList.jsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
+import React from "react";
+import Task from "./Task";
+
+export default function TaskList({ tasks, toggleTask }) {
+  return (
+    <div className="task-list">
+      <div className="lefttasks">
+        {tasks.filter((task) => !task.complete).length} Aufgaben übrig
```
Comment (macfrei) [inline components/calendar_components/TaskList.jsx]: I would probably add the filter as a variable before the return statement for better readabitlity:
```suggestion
  const uncompletedTasks = tasks.filter((task) => !task.complete).length

  return (
    <div className="task-list">
      <div className="lefttasks">
        {uncompletedTasks} Aufgaben übrig
```
Also, what happens if you don't have any tasks yet? And what is the default value if you don't have any `uncompleted` tasks?

Comment (mpagels) [inline components/calendar_components/TaskList.jsx]: I would recommend putting this operation before the return statement and saving it in an menaingful variable. Just the filter function here is not good readable.
 ```js
const numberOfUncompletedTask = tasks.filter((task) => !task.complete).length
```

File: components/calendar_components/TaskList.jsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+      </div>
+      <div className="title_tasklist">
+        <h2>To-Do-List</h2>
+      </div>
+      <div className="back">
+        <div className="value_task">
+          {tasks.map((task) => {
+            return <Task key={task.id} toggleTask={toggleTask} task={task} />;
+          })}
+        </div>
```
Comment (mpagels) [inline components/calendar_components/TaskList.jsx]: Since your component is called `TaskList` and this here looks like a list, you should use the semantic elements `ul` and `li` for creating a list.

Comment (macfrei) [inline components/calendar_components/TaskList.jsx]: Technically this could/should be a semantic `ul` list with `li` children. Just something you might want to think about 🙂

File: components/calendar_components/TaskList.jsx [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        {tasks.filter((task) => !task.complete).length} Aufgaben übrig
+      </div>
+      <div className="title_tasklist">
+        <h2>To-Do-List</h2>
+      </div>
+      <div className="back">
+        <div className="value_task">
+          {tasks.map((task) => {
+            return <Task key={task.id} toggleTask={toggleTask} task={task} />;
+          })}
```
Comment (macfrei) [inline components/calendar_components/TaskList.jsx]: I usually prefer to write the return of the `map` without the explicit `return` within a `return` statement. Just personal preference though.
```suggestion
          {tasks.map((task) =>  (
           <Task key={task.id} toggleTask={toggleTask} task={task} />;
          ))}
```

File: header&footer/header/header.jsx [hunk 9]
```diff
@@ -1,10 +1,10 @@
...
+import React from "react";
+import styles from "../header/header.css";
 
 const Header = () => (
-    <div className= "header">
-        <p>Menu</p>
-    </div>
+  <div className="header">
+    <p>Menu</p>
+  </div>
```
Comment (macfrei) [inline header&footer/header/header.jsx]:
```suggestion
  <header className="header">
    <p>Menu</p>
  </header>
```

File: pages/about.jsx [hunk 10]
```diff
@@ -4,23 +4,23 @@ const About = () => {
...
-          englischen Wort "Identify" (deu.: "identifizieren"). Die Anwendung
-          wurde im Rahmen des Web-Development Bootcamps von "neuefische"
-          programmiert. Die Programmierung dieser Anwendung betrug insgesamt 
-         vier Wochen und dient als nützliche Hilfestellung für den
-          Haushaltsalltag. Die primären Komponenten bestehen aus einem
-          Kassenbuch und einer Einkaufsliste. Als weitere nützliche Features
-          wurden noch ein Taschenrechner mit einfacher Funktion und ein Kalender
-          für die täglichen "To-Do's" einprogramiert. Erweiterungen der
-          Anwendung sind geplant.
+          <strong>HAWI.FY</strong> ist eine Kombination aus den Worten
```
Comment (macfrei) [inline pages/about.jsx:14]: If you use the `strong` to only display it in bold, then I would recommend using a `span` and style it accordingly. The `strong` elements semantic meaning is about conveying an importance. You can read about it here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strong

File: pages/calculator.jsx [hunk 11]
```diff
@@ -1,15 +1,43 @@
 import { Link } from "react-router-dom";
-
+import { useState } from "react";
 
 const Calculator = () => {
```
Comment (macfrei) [inline pages/calculator.jsx:5]: Since this is a component, I would rename the file to `Calculator`, so with an uppercase letter starting.

File: header&footer/header/header.jsx [hunk 12]
```diff
@@ -1,10 +1,10 @@
-import React from 'react';
-import styles from '../header/header.css';
+import React from "react";
+import styles from "../header/header.css";
 
 const Header = () => (
-    <div className= "header">
-        <p>Menu</p>
-    </div>
+  <div className="header">
```
Comment (mpagels) [inline header&footer/header/header.jsx]: Why not using the semantic `header` or headline elements like `h2`?

File: pages/calculator.jsx [hunk 13]
```diff
@@ -1,15 +1,43 @@
...
+  return (
+    <>
+      <Link to="/" className="start">
+        Startseite
+      </Link>
 
-<Link to="/" className="start">Home</Link>
+      <div className="container">
+        <form>
+          <input type="text" value={result} />
```
Comment (macfrei) [inline pages/calculator.jsx:42]: This is not really done yet as far as I can tell, however, please add a `label` element 🙂

File: pages/calender.jsx [hunk 14]
```diff
@@ -1,114 +1,62 @@
...
+
+  function handleClearTasks() {
+    const newTasks = tasks.filter((task) => !task.complete);
+    setTasks(newTasks);
+  }
+
+  function handleAddTask(e) {
+    const name = taskNameRef.current.value;
+    if (name === "") return;
+    console.log(name);
```
Comment (macfrei) [inline pages/calender.jsx:35]: Please remove the `console.log`

File: navbar/Menu.jsx [hunk 15]
```diff
@@ -1,34 +1,39 @@
...
+      <ul className="nav-list">
         <li className="cashbook">
-          <Link to='/income' className="nav-item">Kassenbuch</Link>
+          <Link to="/income" className="nav-item">
+            Kassenbuch
+          </Link>
         </li>
         <li className="shopping">
-          <Link to='/shop' className="nav-item">Einkaufsliste</Link>
+          <Link to="/shop" className="nav-item">
```
Comment (mpagels) [inline navbar/Menu.jsx:15]: This is just a personal idea, why not using `shopping-list` instead of `shop`? When I read `shop` I'm not thinking of an `shopping-list`. ;-)

File: pages/calender.jsx [hunk 16]
```diff
@@ -1,114 +1,62 @@
...
-          </div>
-        </div>
-      </div>
-
-      <div className="todo_field">
-        <h2>To-Do Liste</h2>
-      </div>
+      <TaskList tasks={tasks} toggleTask={toggleTask} />
 
       <input
```
Comment (macfrei) [inline pages/calender.jsx:50]: Where is the `label` element?

File: pages/calender.jsx [hunk 17]
```diff
@@ -1,114 +1,62 @@
...
+        ref={taskNameRef}
         type="form"
```
Comment (macfrei) [inline pages/calender.jsx:52]: I don't think there is a `type="form"` for the `input` element, no?

File: pages/calender.jsx [hunk 18]
```diff
@@ -1,114 +1,62 @@
...
-        </div>
-      </div>
-
-      <div className="todo_field">
-        <h2>To-Do Liste</h2>
-      </div>
+      <TaskList tasks={tasks} toggleTask={toggleTask} />
 
       <input
+        ref={taskNameRef}
```
Comment (macfrei) [inline pages/calender.jsx:51]: Why not use the `onChange` event and add a state for the input value? That way you don't need the `ref`.

File: src/App.css [hunk 19]
```diff
@@ -126,81 +124,122 @@ body {
...
+  font-family: Poetsen One;
+  font-size: 22px;
+}
+
+.back {
+  background-color: #f3e779;
+  height: 100%;
+}
+
+/* Calculator Component */
```
Comment (macfrei) [inline src/App.css:245]: Why not have an extra CSS file for each component? Having all of your CSS in one file might be difficult to navigate and maintain later on.

**Review by macfrei — COMMENTED** [review]
Nice job 👍 I added some suggestions you might want to look at. Keep up the great work 🙂

File: pages/calculator.jsx [hunk 20]
```diff
@@ -1,15 +1,43 @@
...
-    )
-}
+      </div>
+    </>
+  );
+};
 
-export default Calculator;
\ No newline at end of file
+export default Calculator;
```
Comment (mpagels) [inline pages/calculator.jsx]: Would delete that or move it into another pull request.

File: pages/calender.jsx [hunk 21]
```diff
@@ -1,114 +1,62 @@
 import { Link } from "react-router-dom";
+import { useState, useRef, useEffect } from "react";
+import TaskList from "../components/calendar_components/TaskList";
+import { v4 as uuid } from "uuid";
 
 const ToDoList = () => {
+  const [tasks, setTasks] = useState([]);
+  const taskNameRef = useRef();
+  const LOCAL_STORAGE_KEY = "taskApp.tasks";
```
Comment (mpagels) [inline pages/calender.jsx:9]: Would move that to line 5. You don't need that every time created when the component is re-rendered.

File: pages/calender.jsx [hunk 22]
```diff
@@ -1,114 +1,62 @@
...
+
+  useEffect(() => {
+    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(tasks));
+  }, [tasks]);
+
+  function toggleTask(id) {
+    const newTasks = [...tasks];
+    const task = newTasks.find((task) => task.id === id);
+    task.complete = !task.complete;
+    setTasks(newTasks);
```
Comment (mpagels) [inline pages/calender.jsx:24]: Very good making a copy first and then doing the `find()` method.

Another solution would be: (But just FYI) 
```js
const newTask = tasks.map(task => {
    if (task.id === id) {
        task.complete = !task.complet
        return task 
    } else {
        return task 
    }
})
```

File: pages/calender.jsx (hunk 14 above)
Comment (mpagels) [inline pages/calender.jsx:35]: Here is a `console.log()` Highly recommend removing it.

File: pages/calender.jsx [hunk 23]
```diff
@@ -1,114 +1,62 @@
...
         className="input_todos"
+        placeholder="Hier eingeben..."
       />
 
       <div className="delete_clear">
-        <button>Aufgabe entfernen</button>
-        <button>Liste löschen</button>
+        <button onClick={handleAddTask}>Aufgabe hinzufügen</button>
+        <button onClick={handleClearTasks}>Aufgabe entfernen</button>
       </div>
```
Comment (mpagels) [inline pages/calender.jsx:60]: This works, yes, but not sure if a `form` element would be better?

File: pages/income.jsx [hunk 24]
```diff
@@ -1,14 +1,15 @@
...
+    <div className="expense">
+      <Link to="/" className="start">
+        Startseite
+      </Link>
+
+      <h2 className="text">This is the Expense Tracker Site!</h2>
+    </div>
+  );
+};
+
```
Comment (mpagels) [inline pages/income.jsx:14]: This is just an empty site with no value for the user. Would remove it from the pull request and put it back in another story.

File: pages/shoppingList.jsx [hunk 25]
```diff
@@ -1,15 +1,15 @@
...
+  return (
+    <div className="shop">
+      <Link to="/" className="start">
+        Startseite
+      </Link>
+
+      <h2 className="text">This is the Shopping Site!</h2>
+    </div>
+  );
+};
```
Comment (mpagels) [inline pages/shoppingList.jsx:13]: Here as well. This is just a page with no value for the user.

**Review by mpagels — COMMENTED** [review]
Hey Samir, good job.

Some stuff I want to mention:

## Commit messages
Just `change` as a commit message is not good. Please try to find a meaningful commit message.
Here is a [good article](https://chris.beams.io/posts/git-commit/) about writing good commit messages.

## Things, that are not done yet
If you don't have something finished or ready (like the footer -> empty list, calculator that is not working) then it has no place in a pull request. Think of it like, you want to put changes into production. Something that works for you and the user. The calculator is not done yet.

## Component names
files, that is a component, please Capitalize. (E.g.: footer.jsx, header.jsx, income.jsx, shoppingList)

## Misc
- Why did you install `next.js`??
- You installed `styled-components` -> why not using it? Highly recommend using it here.
- Would be a good idea to put your project on vercel, so that reviewer can visit your site and branch.
